### PR TITLE
vagrant: add darwin support

### PR DESCRIPTION
**Pull request by ryanartecona**
###### Motivation for this change

Add darwin support to vagrant.
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [x] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Vagrant on macOS is distributed as a .dmg installer. Luckily, the
internal contents of that archive resemble that of the .deb we use for
linux. In fact, the similarity is enough that if we move its `embedded`
directory to `opt/vagrant/embedded` and its `bin` to `usr/bin` (and back
again after installation), the derivation's installPhase (which replaces
embedded libs & binaries with those from the package's inputs) can
remain exactly the same between macOS and linux.

### CODE REVIEW

**Comment by ryanartecona** [comment]
Pinging maintainers @lovek323 @globin @jgeerds @kamilchm

File: pkgs/development/tools/vagrant/default.nix [hunk 1]
```diff
@@ -110,5 +129,10 @@ in stdenv.mkDerivation rec {
   postFixup = ''
     chmod +x "$out/opt/vagrant/embedded/gems/gems/bundler-1.12.5/lib/bundler/templates/Executable"
     chmod +x "$out/opt/vagrant/embedded/gems/gems/vagrant-$version/plugins/provisioners/salt/bootstrap-salt.sh"
-  '';
+  '' +
+  (stdenv.lib.optionalString stdenv.isDarwin ''
```
Comment (LnL7) [inline pkgs/development/tools/vagrant/default.nix]: Why is this needed? Can't we keep this in the same place as other platforms?

**Comment by LnL7** [comment]
It looks like there's still an issue with the ruby load paths.

```
$ result/bin/vagrant --help
/nix/store/1wmp1mbsxwlkyipjmyvw1lzca7wz4fhv-ruby-2.2.5/lib/ruby/site_ruby/2.2.0/rubygems/core_ext/kernel_require.rb:55:in `require': cannot load such file -- bundler (LoadError)
```

**Comment by ryanartecona** [comment]
@LnL7 It's working for me, even with an almost empty environment (`env -i HOME=$HOME result/bin/vagrant --help` works). Do you mind pasting the output of `env VAGRANT_DEBUG_LAUNCHER=1 result/bin/vagrant --help`? (Warning: that thing will print out all of your current env vars along with its own debug info.)

**Comment by LnL7** [comment]
https://gist.github.com/LnL7/f265ae3de7d0885ae60efc1df5331faf

**Comment by peterhoeg** [comment]
@LnL7, does this file exist 
```/nix/store/hpyq4ir3gx8krr6pnmdxjnd293w2fvxk-vagrant-1.8.6/embedded/gems/gems/vagrant-share-1.1.5/lib/vagrant/pre-rubygems.rb``` ?

My guess is that line 6 in the gist has something to do with it (bad gempath).

**Comment by dipinhora** [comment]
It worked for me on an OS X 10.10.5 machine. Not sure why I don't see the same error as @LnL7  but I built using the following command after cherry-picking the fix onto the current master: `NIX_PATH=nixpkgs=/Users/dipinhora/nixpkgs/ nix-shell --pure --keep-failed  -p vagrant`

In case it helps, the output of `env VAGRANT_DEBUG_LAUNCHER=1 vagrant --help` can be found at: https://gist.github.com/dipinhora/4e898968b20e1155fcc78867131d68c1

**Comment by ryanartecona** [comment]
@LnL7 What do you get if you run this?

```bash
result/embedded/bin/ruby --debug result/embedded/gems/gems/vagrant-1.8.6/lib/vagrant/pre-rubygems.rb --help
```

**Comment by ryanartecona** [comment]
Here's mine for referene: https://gist.github.com/ryanartecona/3d7677aa8b0c7af3b4f5e8d192280deb

**Comment by LnL7** [comment]
@ryanartecona Ok it looks like there might be something strange going on with my laptop, I don't see this issue on my other machine and I have the same issue with rake for example.
Don't understand what's causing this however.

https://gist.github.com/LnL7/a127807bfc4ad88cbf8c267199cb6d05

**Comment by ryanartecona** [comment]
@LnL7 Weird. Ok, if I look in `result/embedded/gems/gems/vagrant-1.8.6/lib/vagrant/pre-rubygems.rb` on line 19 (from your pasted error) on my machine, it's a line with `require_relative "bundler"`. Can you verify the file `result/embedded/gems/gems/vagrant-1.8.6/lib/vagrant/bundler.rb` exists for you?

**Comment by LnL7** [comment]
Yes, that exists.
I'm pretty sure there's just something strange going on with that machine, one last thing I can think of is running using `dtruss` or `opensnoop` to check where your machine finds bundler.

**Comment by ryanartecona** [comment]
@LnL7 Ok. Still, weird.

Definitely not certain I'm doing this right, but my attempt at dtrussing the thing failed. If I `sudo dtruss result/bin/vagrant`, it only tells me what the launcher is doing, even if I pass `-f` to follow children. If I instead `sudo dtruss -f -t open  result/embedded/bin/ruby result/embedded/gems/gems/vagrant-1.8.6/lib/vagrant/pre-rubygems.rb --help`, I get some stuff that looks relevant, but nothing about bundler loading, except for an error that gets printed before any of the dtruss output starts. [gist here](https://gist.github.com/ryanartecona/3830e20a289e2d07d8525b9e9e174fbe).
